### DISCUSSION TITLE
Decoupling logic from the ui

### DISCUSSION THREAD

**Discussion by szsascha** [codefori/vscode-ibmi#230] [Ideas]
Code for IBM i is getting bigger and bigger. It's becoming increasingly difficult to find the right places in the code. Maybe now would be the right time to rethink the current code structure and do a refactoring before adding more stuff.

There are many functions which are not directly depending on the ui. For example the whole IFS handling for down- and uploading files or much more in this direction.

A possible solution could be to take a big part of the logic out of the extension and create a new library out of it. The extension will use this library. But it will bring a bit more structure to the codebase and the library can be also used by other projects.

What's your opinion on this?

**Comment by worksofliam** [comment]
You're not wrong, the extension is getting very big. I'd be curious to how this separate library might look - that's a lot of work in itself.

Of course, like you, have a day job. Finding the time to do all this work is not simple, esp. considering we have a few PRs in the works too.

Of course, I am totally open to PRs of any refactorings.

**Reply by worksofliam** [reply]
@szsascha I slept on it and a clean-up is likely a good idea.

Are you able to bullet point a list of your ideas for what this could look like? I'd be happy to make a start after I have reviewed your PRs.

**Reply by szsascha** [reply]
@worksofliam Thanks ;)

Unfortunately I'm not so into JS architecture and best practices for code structure (are there any at all in JS? ;) )
But the following list came to my mind after I was working on my last 2 PR's:

1. Separate logic and "view" stuff
2. Reduce code redundancies by creating common functions
3. Use a more consistent name convention. For example: All files in api are beginning with uppercase and in views with lowercase
4. Switch to TDD / unit testing for logic functions
5. Try to create smaller methods (not too small!)
6. With the refactoring it could make sense to export all constants into a file and start localizing the extension
7. Rework SQL execution. It's not so nice to have a condition before each runSQL because it is likely not supported on the server.
8. From my point of view I would migrate the whole project to TypeScript because type-safety brings a lot more structure by design into the whole thing. But that's just my opinion. 

So I started this discussion and not as an issue to see if someone else has more points to go.

**Reply by worksofliam** [reply]
Regarding some specific points:

1: yes... I plan to split out the file systems and their corresponding views in a PR over the next week. This is the main place I see this happening.
4: I have been considering this for a while. It should be simple enough to add unit tests to our code base, since it is mostly made up of easily testable code.
7: if you can show me a specific place this is happening I will happily take a look
8: this might have been possible early on in the project. It's grown a ton now and the investment doesn't seem worth it right now.

**Reply by szsascha** [reply]
7: IBMiContent getMemberList for example
8: Yeah that's totally true. This decision needs to be taken as early as possible. Afterwards it's possible to migrate part for part to TS. For example while refactoring. But it's absolutely understandable if you want to focus on other things instead. But there is a possibility that this topic comes back in the future.

**Reply by worksofliam** [reply]
@szsascha This isn't perfect, but I started by splitting up the views folder. I moved the file systems into their own folder.

**Reply by szsascha** [reply]
@worksofliam I've been thinking about how the code quality can be improved with a few simple changes.

What's your opinion on creating a kind of "code convention manifest" (maybe in a CONTRIBUTING.md?) and a new point on the PR checklist to check that the code ist conform to that? A part of the code convention manifest could be the beginning of filenames always with an uppercase first letter for example.

Adding unit test's could also be a part of the PR checklist.

**Reply by szsascha** [reply]
Maybe one more thing to cleanup: Are the /media folder an its contents still needed?

**Reply by szsascha** [reply]
@worksofliam 
In the last days I did some more code work on this extension.

Now I would add a refactoring of the if statements to the list. There are pretty long if blocks. That leads to unreadable code. It can be easily fixed by negative checks and returns.

Just a random example:
<img width="680" alt="image" src="https://user-images.githubusercontent.com/43013380/133894570-a05610fc-aaa8-457f-afb2-da37f8bb76ee.png">

This if block is pretty long. Why not checking !command and then return? That would make it a bit easier to read. But it's just a thing for the future or a refactoring session. Just writing it here to remember it.

**Reply by priceaj** [reply]
@worksofliam Speaking of scalability, would you be open to switching from JS to TypeScript? (if someone submitted a PR?)

Since Typescript is a superset of JS it should be relatively easy to convert, and you can add the types and typescript features later.

Strong typing helps to mitigate alot of errors.

**Reply by worksofliam** [reply]
@priceaj I am open to the idea.

Curious to how big your PR will be. We'll likely need to do a bunch of testing to make sure things continue to work (including our deployment)

**Reply by priceaj** [reply]
@worksofliam Okay so I've done some preliminary testing here https://github.com/priceaj/vscode-ibmi/tree/Typescript

I've moved from npm to Yarn, and added the typescript stuff in, you should basically be able to convert the js file to .ts and it should just work and get the benefit of the strong typing (imagine RPG where type errors are caught by the compiler, not during execution!) 

However I've had a few problems converting some of the files mainly because the code needs some work.

There is also no way I can ever get this merged in at present because it would require renaming every js file which would conflict with every PR. 

What we could do is adopt a progressive strategy, e.g. put the TS infrastructure in place then migrate to TS as we go along by just renaming the file and fixing any type issues that come up

**Reply by szsascha** [reply]
@priceaj 

Sounds great!

Why the move from npm to Yarn?
(Just asking because I didn't know the exact advantages. Never used it before.)

**Reply by priceaj** [reply]
It will ensure consistency between machines because it stores versions and it caches packages... But I mainly use it because it's faster because it does things like parallel package downloads and installs!

**Reply by worksofliam** [reply]
> What we could do is adopt a progressive strategy, e.g. put the TS infrastructure in place then migrate to TS as we go along by just renaming the file and fixing any type issues that come up

@priceaj I think this is the best option for the time being. It's likely the best thing to do, then any future PRs will require the change to TS.
